### PR TITLE
Add TE MXFP8 grouped MoE inference backend

### PR DESCRIPTION
**Pull request by wdykas**
We are breaking this one up. There is like way to many separate things going on here. 

- [] I, the PR author, have personally reviewed every line of this PR.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [20ea711](https://github.com/NVIDIA/Megatron-LM/commit/20ea7118338a818e6ad2809b7f5dac61c2020daa)

**Comment by wdykas** [comment]
Batch-invariant MXFP8 validation for `00cd446d8`:

- Native TE FC1/FC2 now run in deterministic 256-row expert chunks with explicit zero padding and stable expert group indices.
- Exact CUDA-graph test passes for both discrete and single-grouped TE weight layouts across 240- and 320-token buckets; selected log-probability max diff is exactly `0.0`.
- Real EP=4 NVLS inference MoE passes small-vs-large co-batches after adding 96 tokens per rank; shared outputs and selected log probabilities are bitwise equal.
- MXFP8 refit passes all four discrete/single-grouped source-destination combinations while preserving cached graph addresses.
- Local suites: `166 passed, 1 skipped` for MXFP8/refit; `58 passed, 3 skipped` per rank for the full 4x GB200 dispatch/routing suite; Black, isort, pylint, Ruff, py_compile, and diff checks pass.

CI restarted under `Run functional tests`: https://github.com/NVIDIA/Megatron-LM/actions/runs/33902819148

**Comment by wdykas** [comment]
/ok to test [8e8089b](https://github.com/NVIDIA/Megatron-LM/commit/8e8089b60c4b1d62a385263e436b8e80405ffb14)

**Comment by wdykas** [comment]
Performance/numerics update for `3e6aee7ab806262a3411acb303aa71344e75df53`:

The batch-invariant TE path now permutes directly into deterministic 256-row chunk-major expert slots in one Triton kernel, bypassing the previous atomic permutation plus eager PyTorch reorder.

GB200 SM100, PyTorch 2.11.0+cu130, CUDA 13.0, TE 2.18.0+27486e03; H=2688, FFN=1856, 8 local experts, top-k=6. Median complete local-MoE CUDA-graph latency (ms):

| Tokens | Torch MXFP8 | FlashInfer | regular TE | optimized TE BI | old TE BI |
|---:|---:|---:|---:|---:|---:|
| 1 | 0.05952 | unsupported | 0.10861 | 0.09223 | 0.19994 |
| 32 | 0.07175 | unsupported | 0.12452 | 0.08551 | 0.17983 |
| 128 | 0.08214 | 0.04977 | 0.13682 | 0.08690 | 0.18084 |
| 256 | 0.10359 | 0.05629 | 0.13860 | 0.08703 | 0.18909 |
| 384 | 0.13593 | 0.06416 | 0.20645 | 0.15721 | 0.31842 |
| 512 | 0.13996 | 0.06618 | 0.20618 | 0.15597 | 0.33168 |

This is a 2.03–2.17x speedup over the old TE BI path. At 384 tokens, CUDA events fall from 97 to 27; the new direct permutation itself is 6.2 us. Discrete and single-grouped TE layouts are effectively equal (0.15721 vs 0.15549 ms at 384 tokens; 0.15597 vs 0.15490 ms at 512).

Validation on the exact head:

- Exact output and selected-log-prob difference is 0.0 across the 256-token boundary for both discrete and single-grouped TE weights.
- Full local permutation/MXFP8/refit suites: 1035 passed, 1 skipped.
- Refit passes all four discrete/single-grouped source-destination combinations while preserving storage.
- Full 4x GB200 NVLS subset: 30 passed, 3 skipped per rank.
- Repository CI lint command passes at 10.00/10 after fixing the branch-initialization diagnostics.

Functional CI restarted here: https://github.com/NVIDIA/Megatron-LM/actions/runs/34618836809 (lint green; GPU matrix currently queued).

File: megatron/core/inference/moe/fused_moe.py [hunk 1]
```diff
@@ -74,6 +128,363 @@ def _mxfp8_grouped_mm(act: MXFP8Tensor, weight: MXFP8Tensor, offs: torch.Tensor)
...
+    """Launch grouped TE GEMM without inheriting te_native's starved workspace.
+
+    The te_native batch-invariant backend restricts ordinary TE GEMMs to a 1 KiB
+    workspace to disqualify split-K algorithms. TE's device-metadata grouped
+    kernel requires its normal cuBLASLt workspace even when every GEMM has a
+    fixed M. Recover the unrestricted size and call the device-metadata API
+    directly for this fixed-shape path only.
+    """
+    sm_count = _get_te_sm_count()
+    sm_count -= int(os.getenv("NVTE_EXT_MARGIN_SM", str(sm_count)))
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/fused_moe.py:328]: Verified against the exact pinned Transformer Engine implementation: its Python wrapper computes `sm_count - int(os.getenv("NVTE_EXT_MARGIN_SM", str(sm_count)))`, so an unset variable deliberately passes 0, and TE documents 0 as using cuBLAS heuristics. Changing the default to `"0"` would instead pass the full device SM count and diverge from TE. I kept the implementation and added an explicit regression assertion in 4e418880f. Sources: https://github.com/NVIDIA/TransformerEngine/blob/27486e03cfc1fa41f6932dcecdc47c71c47eac3e/transformer_engine/pytorch/cpp_extensions/gemm.py#L394-L395 and https://github.com/NVIDIA/TransformerEngine/blob/27486e03cfc1fa41f6932dcecdc47c71c47eac3e/transformer_engine/common/include/transformer_engine/gemm.h

File: megatron/core/inference/moe/permute.py [hunk 2]
```diff
@@ -295,6 +295,30 @@ def _permute_tokens_kernel(
...
+    n_used = tl.load(n_used_ptr)
+    if pid >= n_used:
+        return
+    for row in tl.range(pid, max_rows, NUM_BLOCKS):
+        if row < n_used:
+            if tl.load(permutation_map_ptr + row) < 0:
+                for h in tl.range(0, hidden_dim, BLOCK_H):
+                    offsets = h + tl.arange(0, BLOCK_H)
+                    mask = offsets < hidden_dim
+                    tl.store(hidden_ptr + row * hidden_dim + offsets, 0.0, mask=mask)
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/permute.py]: Fixed in 4e418880f by explicitly promoting `row` to `tl.int64` before multiplying by `hidden_dim`. The focused GB200 permutation suite, including the existing >32-bit offset regression, passes.

File: megatron/core/inference/moe/fused_moe.py [hunk 3]
```diff
@@ -74,6 +128,363 @@ def _mxfp8_grouped_mm(act: MXFP8Tensor, weight: MXFP8Tensor, offs: torch.Tensor)
...
+        _te_batch_invariant_grouped_gemm(
+            normalized_weight,
+            grouped_input,
+            grouped_output,
+            alpha,
+            beta,
+            workspace_setup,
+            workspace_cublas,
+        )
+    return output_data
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/inference/moe/fused_moe.py:496]: The 256-row calls are intentional for batch invariance: FC2's TE recipe selection depends on the group index, so coalescing experts/chunks changes the selected kernel as the batch shape changes. The optimized path already reuses one serialized workspace and the PR benchmark measured 2.03–2.17x over the prior BI path (97 to 27 CUDA events at 384 tokens; direct permutation 6.2 us). The focused cross-256-boundary invariance tests pass in both discrete and single-grouped layouts. No code change here because increasing the chunk changes the required determinism semantics.

File: megatron/core/extensions/transformer_engine.py [hunk 4]
```diff
@@ -165,9 +165,10 @@ class TEQuantizationRecipe:
     If an amax reduction is applicable, such as in per-tensor quantization recipe,
     whether to reduce only along TP groups.
     """
-    fp8_param: bool = False
+    fp8_param: Optional[bool] = None
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/extensions/transformer_engine.py:168]: Implemented suggested option (a) in 4e418880f: `parse_from_config` emits a targeted one-time `UserWarning` when an FP8 per-module recipe omits `fp8_param`, explaining the new inheritance behavior and how to request the old forced-BF16 behavior explicitly. Added a regression test; I am also adding the compatibility note to the PR description.

File: megatron/core/transformer/moe/experts.py [hunk 5]
```diff
@@ -1331,6 +1341,58 @@ def refresh_flashinfer_mxfp8_weights(self) -> bool:
             prepare_routed_mxfp8_weights(canonical_weight, out=routed_weight)
         return True
 
+    @torch.inference_mode(False)
+    @torch.no_grad()
+    def _build_te_mxfp8_weights(self):
+        """Compatibility entry point for callers that require native TE MXFP8 storage."""
+        InferenceGroupedMLP._build_te_inference_weights(self, expected_mxfp8=True)
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/moe/experts.py]: Removed the dead `_build_te_mxfp8_weights` wrapper in 4e418880f and updated its only test caller to use `_build_te_inference_weights(..., expected_mxfp8=True)` directly.

File: megatron/core/fp8_utils.py [hunk 6]
```diff
@@ -178,16 +178,27 @@ def get_grouped_quantized_members(
...
             raise RuntimeError(
-                "Grouped quantized parameter is missing cached member tensors. "
+                "Grouped parameter is missing cached member tensors. "
                 "Create them outside the training critical path."
             )
-        quantized_members = grouped_tensor.split_into_quantized_tensors()
-        grouped_tensor.quantized_tensors = quantized_members
-    return quantized_members
+        members = grouped_tensor.split_into_quantized_tensors()
+        grouped_tensor.quantized_tensors = members
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/fp8_utils.py:207]: Kept the name because this is Transformer Engine's own cache/API convention: TE exposes `quantized_tensors` and `split_into_quantized_tensors` for grouped members even when the grouped storage is high precision. Added a docstring in 4e418880f explaining that the helper deliberately shares TE's stable member-view cache.

File: megatron/core/transformer/transformer_config.py [hunk 7]
```diff
@@ -1784,7 +1805,44 @@ def __post_init__(self):
...
+                )
+            for pattern in mxfp8_parameter_filters:
+                if pattern is None:
+                    continue
+                try:
+                    re.compile(pattern)
+                except re.error as error:
+                    raise ValueError(
+                        f"Invalid MXFP8 parameter regex {pattern!r}: {error}"
+                    ) from error
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/transformer_config.py:1833]: Removed the duplicate utility-layer regex compilation/validation in 4e418880f. `TransformerConfig` is now the single validation owner before configured conversion begins.

File: megatron/core/transformer/transformer_config.py [hunk 8]
```diff
@@ -3507,23 +3569,71 @@ def _scope_to_str(s):
...
+                    and self.inference_grouped_gemm_backend == InferenceGroupedGemmBackend.TORCH
+                    and bool(self.fp8)
+                    and self.fp8_recipe == Fp8Recipe.mxfp8
+                    and self.fp8_param
+                    and not self.fp4
+                    and (
+                        (not self.gated_linear_unit and self.activation_func == squared_relu)
+                        or (self.gated_linear_unit and self.activation_func == F.silu)
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/transformer/transformer_config.py]: Refactored in 4e418880f: the common MXFP8-parameter predicate and activation predicates are computed once, then backend-specific support is expressed through one backend map.

File: megatron/core/resharding/utils.py [hunk 9]
```diff
@@ -288,9 +290,35 @@ def named_refit_tensors(module: torch.nn.Module):
...
+    metadata_attributes = (
+        'allreduce',
+        'tensor_model_parallel',
+        'partition_dim',
+        'partition_stride',
+        'partition_sizes',
+        'is_gtp_weight_remat',
+        'group',
+        'pad_length',
+    )
```
[the remark this answers is not in the document]
Comment (wdykas) [reply megatron/core/resharding/utils.py]: Centralized the superset in a dependency-free `megatron.core.parameter_metadata.PARAMETER_SHARDING_ATTRIBUTES` constant in 4e418880f and reused it from both replacement and refit member-view paths. Keeping the module dependency-free avoids a cold-import cycle through tensor-parallel quantization utilities.

**Comment by wdykas** [comment]
Validated the current head `34a51b187ff8922e56efdad49df99983e421b610` end to end with NVIDIA-NeMo/RL#3208 plus the stacked NeMo fix in https://github.com/YigongQin/RL/pull/3.

Production-shaped configuration:

- Qwen3-30B-A3B
- colocated Megatron-Inference, TP=1 / EP=4, NVLS dispatcher
- native TE grouped MXFP8 only for routed expert FC1/FC2
- first 2 and last 4 layers kept in BF16

Exact selected-token comparison (generation logprobs versus training-policy logprobs):

```json
{
  "selected": 32,
  "mismatches": 0,
  "max_abs_diff": 0,
  "exact_equal": true
}
```

The one-step NeMo run exited 0 and reported Generation KL Error 0.0000. The persisted comparison artifact had SHA-256 `233d0a2059aa453ab20a1568f1dafe0cd8b3c09b26cf056e03e20ab8835b3155`.

Focused validation also passed:

- 34 distributed MCore GPU tests, including TE grouped-GEMM parity, both TE MXFP8 weight formats, refit refresh, NVLS MoE parity, and Qwen3 dynamic-engine parity
- 27 affected MCore tests after final formatting
- 6 focused NeMo unit tests for TP=1 logprob exactness and dedicated/colocated provider configuration
